common/src/main/java/org/apache/celeborn/common/network/server/memory/BufferQueue.java

```diff
@@ -0,0 +1,166 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.celeborn.common.network.server.memory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.annotation.Nullable;
+
+import io.netty.buffer.ByteBuf;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+// Assume that max-managed memory for a MapDataPartition is (2^31 * buffersize)
+public class BufferQueue {
+  public static final Logger logger = LoggerFactory.getLogger(BufferQueue.class);
+
+  private final Queue<ByteBuf> buffers = new ConcurrentLinkedQueue<>();
+
+  private final MemoryManager memoryManager = MemoryManager.instance();
+
+  /** Number of buffers occupied by this buffer queue (added but still not recycled). */
+  private final AtomicInteger numBuffersOccupied = new AtomicInteger();
+
+  private final AtomicInteger pendingRequestBuffers = new AtomicInteger();
+
+  /** Whether this buffer queue is released or not. */
+  private volatile boolean isReleased = false;
+
+  private volatile int localBuffersTarget = 0;
+
+  /** Returns the number of available buffers in this buffer queue. */
+  public int size() {
+    return buffers.size();
+  }
+
+  /**
+   * Returns an available buffer from this buffer queue or returns null if no buffer is available
+   * currently.
+   */
+  @Nullable
+  public ByteBuf poll() {
+    return buffers.poll();
+  }
+
+  /**
+   * Add buffers and increment numBufferOccupied. Free all buffers to global memory pool if this
+   * buffer queue is released.
+   */
+  public synchronized void add(Collection<ByteBuf> availableBuffers) {
+    if (!isReleased) {
+      buffers.addAll(availableBuffers);
+      numBuffersOccupied.addAndGet(availableBuffers.size());
+      pendingRequestBuffers.addAndGet(-1 * availableBuffers.size());
+    } else {
+      for (ByteBuf availableBuffer : availableBuffers) {
+        memoryManager.recycleReadBuffer(availableBuffer);
+      }
+    }
+  }
+
+  public void recycle(ByteBuf buffer) {
+    if (isReleased || numBuffersOccupied.get() > localBuffersTarget) {
+      recycleToGlobalPool(buffer);
+    } else {
+      recycleToLocalPool(buffer);
+    }
+  }
+
+  public synchronized void recycleToGlobalPool(ByteBuf buffer) {
+    numBuffersOccupied.decrementAndGet();
+    memoryManager.recycleReadBuffer(buffer);
+  }
+
+  public synchronized void recycleToLocalPool(ByteBuf buffer) {
```

Comment:
IMO it is inefficient to synchronize ```recycleToLocalPool``` and ```recycleToGlobalPool```, especially ```recycleToLocalPool```, because it's high-frequent operation